### DISCUSSION TITLE
can I set all node label in the bottom?

### DISCUSSION THREAD

**Discussion by FFengIll** [terrastruct/d2#216] [General]
I have tested d2 with my figures, and I am confused I **can not control the node label position** (specially for `shape: image`).

In d2 output, label set in bottom / right side and top which follows the layout strategy automatically.

Can we have any possible to force all label in the top or in the bottom to confirm consistence for some figures?

> the official case is just a good example
> ![](https://raw.githubusercontent.com/terrastruct/d2/master/docs/examples/twitter/out.png)

**Comment by alixander** [comment]
are you using dagre/elk layout engines? If so, there's a ticket to track automatic layout placement so you won't need to specify. https://github.com/terrastruct/d2/issues/203

**Comment by FFengIll** [comment]
@alixander  thanks for your replay.
The automatic layout is a good feature in d2, and future related features must be exciting.

But my core question is `how to set alll icon labels on the same side` (aka. `shape:image icon: xxx` with label).
In most cases (maybe for myself only), the general node label can be free, but the `icon only node` must have a consistent label position (usually, just bellow the icon).

> you could see the image in issue - `twitter icon` labeled on the `left`, and `service icon` labeld on the `top`.

**Comment by alixander** [comment]
ah i see. in that case, this is indeed a different issue. 

i'd be curious to know if this is generally desirable. E.g. if you had arrows coming into icon x from the bottom, wouldn't you want that icon's label to be positioned at the top?

**Comment by FFengIll** [comment]
> ah i see. in that case, this is indeed a different issue.
> 
> i'd be curious to know if this is generally desirable. E.g. if you had arrows coming into icon x from the bottom, wouldn't you want that icon's label to be positioned at the top?

in my case, the answer is **no**.
no mater which node slot linked by an edge, the `icon only node` always has a label on the bottom.

I think my issue may not general enough, and I want to do some discuss if possible.

---

In fact, I have collected a group of figures as ref in my daily work , and many of them use the style as I say, so I feel strage when the icon  label place in different sides (especially in one figure).

**Comment by alixander** [comment]
I'll let other people chime in. I've heard this feedback from others too actually but I find it hard to believe that it's desirable. 

Dynamic labels:
<img width="212" alt="Screen Shot 2022-11-25 at 5 33 02 PM" src="https://user-images.githubusercontent.com/3120367/204067433-3370e6dd-6fb9-4594-857f-51144e6efc5b.png">

Static labels:
<img width="216" alt="Screen Shot 2022-11-25 at 5 33 08 PM" src="https://user-images.githubusercontent.com/3120367/204067440-800f5b1b-ae25-4725-a6bb-53ae87d30cd7.png">

By the way we also do this for other labels in TALA. E.g. for a container label, if there's arrows coming into it from all sides, the label will try to place inside. If there's stuff in the center, it'll place inside near top center. If there's stuff there, then down center, and so on. 

Likewise for connection labels. It tries to do somewhere near halfway of a connection, but if it hugs another node or connection, or merges, then it'll choose another.

**Comment by FFengIll** [comment]
In design, we usually make `icon only` node as a special type which place its label on the bottom.

Furthermore, we use `icon` and its `label` as an object, so the arrow could not cross through the label.

A diff as bellow
![image](https://user-images.githubusercontent.com/7359555/204067799-f848e857-5f1a-43b4-8754-517f7df566c7.png)

**Comment by alixander** [comment]
ohhhh, yeah, okay, i understand now. kk this might be the best solution actually. we'll do this in this or next release. thank you
